src/sdsc/xslt.py

```diff
@@ -0,0 +1,396 @@
+#
+# Copyright (c) 2015-2016 SUSE Linux GmbH
+#
+# This library is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 2.1 of the License, or (at your option) any later version.
+#
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
+#
+
+from lxml import etree
+import os.path
+
+from .regex import (sentenceends, lastsentenceends, re_compile)
+
+# termcheck
+# buildtermdata
+# dupecheck
+
+
+
+# --------------------------------------------------------------------
+# Helper functions
+# --------------------------------------------------------------------
+def findtagreplacement(text):
+    """Search through a token to see whether it contains a replaced tag.
+    Allows finding a single tag replacement only.
+
+    text - str(): a token
+    """
+
+    tagfound = False
+    tagtype = None
+    tokens = 1
+
+    tagreplacement = re_compile(r'##@(\w+)-(\d+)##')
+    tagsreplaced = tagreplacement.search(text)
+
+    if tagsreplaced:
+        tagfound = True
+        tagtype = str(tagsreplaced.group(1))
+        tokens = int(tagsreplaced.group(2))
+
+    return (tagfound, tagtype, tokens)
+
+
+def removepunctuation(word, start=False, end=False):
+    """Removes punctuation from beginning/end of tokens (i.e. does not
+    expect sentences!).
+
+    Parameters:
+    text  - str(), token that we want to look at
+    start  - bool(): remove punctuation from start of token?
+    apos   - bool(): remove punctuation from end of token?
+    """
+
+    if isinstance(word, list):
+        if start:
+            word = [removepunctuation(word[0], start=True)] + word[1:]
+        if end:
+            word = word[:-1] + [removepunctuation(word[-1], end=True)]
+        return word
+
+    startpunctuation = '([{"\'¡¿'
+    endpunctuation = ')]}/\\"\',:;!?.‥…‼‽⁇⁈⁉'
+
+    if start:
+        word = word.lstrip(startpunctuation)
+    if end:
+        word = word.rstrip(endpunctuation)
+    return word
+
+
+def canBeDupe(word):
+    """TAGIGNORE: character sequences that should be ignored by the duplicate
```

Comment:
"Checks if a word is eligible to be a duplicated word or if it should be ignored because it is actually a tag replacement."